### DISCUSSION TITLE
Remote configuration

### DISCUSSION THREAD

**Discussion by rogeriolino** [symfony/symfony#43045] [Q&A]
## What

Spring Cloud has a pretty good feature called Spring Config Server: https://cloud.spring.io/spring-cloud-config/reference/html/

This config server allow us to set up a remote configuration server using git repository to store all configuration files, and Spring apps loads these configuration at startup time.

## Why

- Centralizing application configuration files
- Allowing to update props/config without rebuilding the application (in case of loading configuration at startup time)
- For bigger projects it's good to control teams access

## How

PHP projects can be categorized in two types of app executions: traditional (php-fpm) and long running (swoole, php-pm, react-php).

So, in the Symfony configuration component, would be great to allow retrieving remote configuration from git repo choosing when to load it: on startup or build time.

How do you guys think about it?

Regards

**Comment by GromNaN** [comment]
I have implemented such mechanism with a combination of Dotenv, hashicorp vault and vault-agent. The configuration that needs to be loaded dynamically is stored in vault. The vault-agent runs on the same server that php-fpm (or a sidecar container on k8s) and generates the file .env.local every 5 minutes.
Having a distinct process to reload the configuration does not impact the performance of the application.
This is very convenient to manage changing secrets, dynamic configuration and basic feature-flags.

It's also possible to write your own implementation of EnvVarLoaderInterface (what I did before going back to the solution described upper)

**Reply by rogeriolino** [reply]
Nice solution. Congrats.
